File: VideoListerLibrary/AuthCredentials.cs
/*
 * SPDX-FileCopyrightText: 2021-2022 The Video Lister Contributors
 *
 * SPDX-License-Identifier: BSD-3-Clause
 */

using System;
using System.IO;
using System.Threading;
using System.Threading.Tasks;
using Google.Apis.Auth.OAuth2;
using Google.Apis.Services;
using Google.Apis.Util.Store;
using Google.Apis.YouTube.v3;

namespace VideoListerLibrary
{
    /// <summary>
    ///   The credentials needed to get personal play list, e.g. liked videos.
    /// </summary>
    public class AuthCredentials
    {
        /// <summary>
        ///   Asynchronously gets YouTube credential using OAuth 2.0.
        /// </summary>
        /// <param name="clientFile">
        ///   A string to the filename of the client secret. The default value
        ///   is "client_secret.json"
        /// </param>
        /// <returns>
        ///   <para>
        ///     The OAuth 2.0 and <see cref="YouTubeService" /> used to get
        ///     video meta-data.
        ///   </para>
        /// </returns>
        /// <exception cref="FileNotFoundException" />
        [STAThread]
        public async Task<YouTubeService> GetAuthCredentials(string clientFile = "client_secret.json")
        {
            UserCredential Credential;
            using (FileStream Stream = new FileStream(
                clientFile,
                FileMode.Open,
                FileAccess.Read))
            {
                Credential = await GoogleWebAuthorizationBroker.AuthorizeAsync(
                    GoogleClientSecrets.FromStream(Stream).Secrets,
                    new[] { YouTubeService.Scope.YoutubeReadonly },
                    "user",
                    CancellationToken.None,
                    new FileDataStore(GetType().ToString()));
            }

            return new YouTubeService(new BaseClientService.Initializer()
            {
                HttpClientInitializer = Credential,
                ApplicationName = GetType().ToString()
            });
        }
    }
}

File: video_lister/auth_credentials.py
# SPDX-FileCopyrightText: 2022 The Video Lister Contributors

# SPDX-License-Identifier: BSD-3-Clause
"""A method to get YouTube user data for public and private playlists.
Requires a client secret data from cloud console."""

from typing import Any

from google_auth_oauthlib.flow import InstalledAppFlow
from googleapiclient.discovery import build


def get_auth_credentials(client_secret: str = "client_secret.json") -> Any:
    """The credentials needed to get personal play list, e.g. liked
    videos.

    Args:
        client_secret (str, optional): A string to the filename of the client
        secret. Defaults to "client_secret.json".

    Returns:
        Any: The OAuth 2.0 used to getvideo meta-data.
    """

    credentials = InstalledAppFlow.from_client_secrets_file(
        client_secret, ["https://www.googleapis.com/auth/youtube.readonly"]
    ).run_console()

    return build("youtube", "v3", credentials=credentials)
